### PR TITLE
Add wrappers for the selectors of skmatter

### PR DESCRIPTION
**Pull request by PicoCentauri**
TensorMap versions for FPS and CUR sample and feature selection. The current implementation works only for a `TensorMap` without components.

# TODO
 - [x] Tests

----
:books: Documentation preview :books:: https://equisolve--39.org.readthedocs.build/en/39/

### CODE REVIEW

**Comment by PicoCentauri** [comment]
@DavideTisi the branch should be usable for the sparse kernel model :-). At least the basic tests I wrote are passing.

**Comment by DavideTisi** [comment]
thanks you are amazing

File: src/equisolve/numpy/_selection.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    Create a selector for each block. The selection will be done based
+    the values of each :py:class:`TensorBlock`.
+    """
+
+    def __init__(
+        self, selector: GreedySelector, selection_type: str, **selector_arguments
+    ) -> None:
+        self.selection_type = selection_type
+        self.selector_arguments = selector_arguments
+        self.selector = selector
```
Comment (Luthaf) [inline src/equisolve/numpy/_selection.py:26]:
```suggestion
        self.selector_class = selector
```

we are not storing the instance here, right?

Comment (PicoCentauri) [reply src/equisolve/numpy/_selection.py:26]: Not an instance, but the class.
The instance per class is also saved.

Comment (agoscinski) [reply src/equisolve/numpy/_selection.py:26]: I think the rename @Luthaf suggested would be also good.

According to the typing it should be an instance, but it is a class. I think the correct typing would be a `type` or to be more specific Type[GreedySelector]
```python
from typing import Type
...
selector: Type[GreedySelector]
```

This is how mypy deals with it
```python
from typing import Type

class A: pass
class B(A): pass
class C: pass

def fooA() -> Type[A]:
    return A # works

def fooB() -> Type[A]:
    return B # works

def fooC() -> Type[A]:
    return C # cause error

```
Even though for GreedySelector some stuff is weird
```python
from skmatter._selection import GreedySelector
from skmatter.feature_selection import FPS
from typing import Type

def foo_correct() -> Type[GreedySelector]:
    return FPS # works

def foo_why_does_this_work() -> Type[GreedySelector]:
    return type(5) # works

def foo_wrong() -> Type[GreedySelector]:
    return 5 # error
```
It seems to default to `Type[Any]`, because there is no typing support in skmatter? Anyway we can think about this another time.

**Review by agoscinski — COMMENTED** [review]
Is there a plan to remove the FPS and CUR, since you can use them with equisolve GreedySelector?

One probably can do a thinner class to allow easier deployment
```python
from equisolve.numpy._selection import GreedySelector
from skmatter._selection import _FPS
class FPS(GreedySelector):
    def __init__(self, selection_type: str, **selector_arguments) -> None:
        super().__init__(FPS, selection_type, **selector_arguments)
```

File: src/equisolve/numpy/_selection.py (hunk 1 above)
Comment (PicoCentauri) [reply src/equisolve/numpy/_selection.py:26]: Yes I did the renaming!

File: src/equisolve/numpy/_selection.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+class GreedySelector:
+    """Wrapper for :py:class:`skmatter._selection.GreedySelector` for a
+    :py:class:equistore.TensorMap`.
+
+    Create a selector for each block. The selection will be done based
+    the values of each :py:class:`TensorBlock`.
+    """
+
+    def __init__(
+        self, selector: GreedySelector, selection_type: str, **selector_arguments
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply src/equisolve/numpy/_selection.py:22]: yes! good point.

File: src/equisolve/numpy/_selection.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            raise ValueError("No selections. Call fit method first.")
+
+        blocks = []
+        for key, block in X:
+            try:
+                selector = self.selectors[key]
+            except KeyError as err:
+                raise ValueError(f"Block with key {key} does not exist.") from err
+
+            mask = selector.get_support()
```
Comment (agoscinski) [inline src/equisolve/numpy/_selection.py:92]: I think we need to store the mask = selector.get_support() into a tensorblock (or the selected idx) somewhere in the fit function to be compatible with torch.nn.Module. I am not sure how you to do this operation with equistore only but maybe you have an idea?
```python
blocks.append(slice_block(block, properties=block.properties[mask]))
```

Comment (Luthaf) [reply src/equisolve/numpy/_selection.py:92]: Sounds related to https://github.com/lab-cosmo/equistore/issues/187

Comment (PicoCentauri) [reply src/equisolve/numpy/_selection.py:92]: > I think we need to store the mask = selector.get_support() into a tensorblock (or the selected idx) somewhere in the fit function to be compatible with torch.nn.Module. I am not sure how you to do this operation with equistore only but maybe you have an idea?
> 
> ```python
> blocks.append(slice_block(block, properties=block.properties[mask]))
> ```

After discussing with @agoscinski  we agreed that we only store the support as List[List[int]]. This allows the direct usage of torch later.





> Sounds related to [lab-cosmo/equistore#187](https://github.com/lab-cosmo/equistore/issues/187)

Yes, a bit. If we store a list of ints we can change the logic of slice block later once we have this functionality.

**Comment by jwa7** [comment]
Hey @PicoCentauri @agoscinski @Luthaf what's the status with this PR? Is there anything that needs doing that I can help with to get this merged?

This PR is useful - particularly short term for me for writing a rascaline / equistore / equisolve / chemiscope cookbook example of sample/feature selection!

Let me know :)

**Comment by agoscinski** [comment]
We are a bit busy with other things at the moment, but I think the branch is mostly done. In two week I'll have a look and merge it. Feel free to remind me if I forget about it.

**Comment by jwa7** [comment]
Ok @PicoCentauri I can work on that change for `slice_block()`
